v2/apiserver/internal/core/kubernetes/substrate.go

```diff
@@ -568,3 +655,622 @@ func (s *substrate) countRunningPods(
+		Name:            jobName, // Primary container takes the job's name
+		Image:           jobSpec.PrimaryContainer.Image,
+		ImagePullPolicy: corev1.PullPolicy(jobSpec.PrimaryContainer.ImagePullPolicy), // nolint: lll
+		Command:         jobSpec.PrimaryContainer.Command,
+		Args:            jobSpec.PrimaryContainer.Arguments,
+		Env:             make([]corev1.EnvVar, len(jobSpec.PrimaryContainer.Environment)), // nolint: lll
+		VolumeMounts:    []corev1.VolumeMount{},
+	}
+	i := 0
+	for key := range jobSpec.PrimaryContainer.Environment {
+		containers[0].Env[i] = corev1.EnvVar{
+			Name: key,
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: fmt.Sprintf("job-%s-%s", event.ID, jobName),
+					},
+					Key: fmt.Sprintf("%s.%s", jobName, key),
+				},
+			},
+		}
+		i++
+	}
+	if jobSpec.PrimaryContainer.UseWorkspace {
+		containers[0].VolumeMounts = []corev1.VolumeMount{
+			{
+				Name:      "workspace",
+				MountPath: jobSpec.PrimaryContainer.WorkspaceMountPath,
+			},
+		}
+	}
+	if jobSpec.PrimaryContainer.UseSource {
+		containers[0].VolumeMounts = append(
+			containers[0].VolumeMounts,
+			corev1.VolumeMount{
+				Name:      "vcs",
+				MountPath: jobSpec.PrimaryContainer.SourceMountPath,
+			},
+		)
+	}
+	if jobSpec.PrimaryContainer.UseHostDockerSocket {
+		containers[0].VolumeMounts = append(
+			containers[0].VolumeMounts,
+			corev1.VolumeMount{
+				Name:      "docker-socket",
+				MountPath: "/var/run/docker.sock",
+			},
+		)
+	}
+	if jobSpec.PrimaryContainer.Privileged {
+		tru := true
+		containers[0].SecurityContext = &corev1.SecurityContext{
+			Privileged: &tru,
+		}
+	}
+
+	// Now add all the sidecars...
+	i = 1
+	for sidecarName, sidecarSpec := range jobSpec.SidecarContainers {
+		containers[i] = corev1.Container{
+			Name:            sidecarName,
+			ImagePullPolicy: corev1.PullPolicy(sidecarSpec.ImagePullPolicy),
+			Command:         sidecarSpec.Command,
+			Args:            sidecarSpec.Arguments,
+			Env:             make([]corev1.EnvVar, len(sidecarSpec.Environment)),
+		}
+		j := 0
+		for key := range sidecarSpec.Environment {
+			containers[i].Env[j] = corev1.EnvVar{
+				Name: key,
+				ValueFrom: &corev1.EnvVarSource{
+					SecretKeyRef: &corev1.SecretKeySelector{
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: fmt.Sprintf("job-%s-%s", event.ID, jobName),
+						},
+						Key: fmt.Sprintf("%s.%s", sidecarName, key),
+					},
+				},
+			}
+			j++
+		}
+		if sidecarSpec.UseWorkspace {
+			containers[i].VolumeMounts = []corev1.VolumeMount{
+				{
+					Name:      "workspace",
+					MountPath: sidecarSpec.WorkspaceMountPath,
+				},
+			}
+		}
+		if sidecarSpec.UseSource {
+			containers[i].VolumeMounts = append(
+				containers[i].VolumeMounts,
+				corev1.VolumeMount{
+					Name:      "vcs",
+					MountPath: sidecarSpec.SourceMountPath,
+				},
+			)
+		}
+		if sidecarSpec.UseHostDockerSocket {
```

Comment:
It feels like adding the specific volumes for sources, workspaces, and Docker sockets is done at least 3 times, IIRC.

Would it make sense to have some simple helper functions that do this?

Reply:
I'll look at it. iirc, I tried that at one point and things were less readable, but it's very reasonable to give that another try.

Reply:
Yes. I can DRY this up nicely.

Thanks for encouraging me to do so.